modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/handler/MediaHandler.java

```diff
@@ -122,22 +124,53 @@ public WebResponse handle( final WebRequest webRequest, final WebResponse webRes
             throw WebException.notFound( "Not a valid media url pattern" );
         }
 
-        String project = matcher.group( "project" );
-        String branch = Objects.requireNonNullElse( matcher.group( "branch" ), "master" );
-        String type = matcher.group( "mediaType" );
-        ContentId id = ContentId.from( matcher.group( "id" ) );
-        String fingerprint = matcher.group( "fingerprint" );
-        String restPath = matcher.group( "restPath" );
+        final RepositoryId repositoryId = RepositoryId.from( ProjectConstants.PROJECT_REPO_ID_PREFIX + matcher.group( "project" ) );
+        final Branch branch = Branch.from( Objects.requireNonNullElse( matcher.group( "branch" ), "master" ) );
+        final String type = matcher.group( "mediaType" );
+        final ContentId id = ContentId.from( matcher.group( "id" ) );
+        final String fingerprint = matcher.group( "fingerprint" );
+        final String restPath = matcher.group( "restPath" );
 
-        RepositoryId repositoryId = RepositoryId.from( ProjectConstants.PROJECT_REPO_ID_PREFIX + project );
+        if ( webRequest.getEndpointPath() != null )
+        {
+            verifyMediaScope( matcher.group( "context" ), repositoryId, branch, webRequest );
+        }
 
-        PortalRequest portalRequest = createPortalRequest( webRequest, repositoryId, Branch.from( branch ) );
+        PortalRequest portalRequest = createPortalRequest( webRequest, repositoryId, branch );
 
         return executeInContext( repositoryId, branch, () -> type.equals( "attachment" )
             ? doHandleAttachment( portalRequest, id, fingerprint, restPath )
             : doHandleImage( portalRequest, id, fingerprint, restPath ) );
     }
 
+    private void verifyMediaScope( final String projectContext, final RepositoryId repositoryId, final Branch branch,
+                                   final WebRequest webRequest )
+    {
+        final String mediaServiceScope = VirtualHostContextHelper.getMediaServiceScope();
+        if ( mediaServiceScope != null )
+        {
+            final Set<String> scopes =
+                Arrays.stream( mediaServiceScope.split( ",", -1 ) ).map( String::trim ).collect( Collectors.toSet() );
+
+            if ( scopes.stream().noneMatch( scope -> scope.equals( projectContext ) ) )
+            {
+                throw WebException.notFound( "Not a valid media url pattern" );
+            }
+        }
+        else
+        {
+            if ( webRequest.getRawPath().startsWith( "/site/" ) || webRequest.getRawPath().startsWith( "/admin/site/" ) )
```

Comment:
check that it is PortalRequest - if not  throw 404
in case it is - use method `PortalRequest.isSiteBase`

Reply:
Great idea to check that this is a `PortalRequest` and that `PortalRequest.isSiteBase` is true. However, if we throw a 404 error, it would imply that it works only for the site, which is incorrect.